### DISCUSSION TITLE
Sensitivity accuracy

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#459] [General]
Hi,

I use the reverse mode AD for calculating the sensitivities.

How do i check the accuracy of my sensitivity? In the DAFoam papers, I believe you use a complex step method as your reference. And in the paper on field inversion, you use the forward mode AD. But when i tried using the forward mode AD, I get a key error when i do that. I looked into the source code, pasting it here (pyDAFoam.py) 
```
# Whether to use AD: Mode options: forward, reverse, or fd. If forward mode AD is used
        ## the seedIndex will be set to compute derivative by running the whole primal solver.
        ## dvName is the name of design variable to set the seed for the forward AD
        ## setting seedIndex to -1 for dFdField will assign seeds for all design variables.
        ## If reverse mode is used, the adjoint will be computed by a Jacobian free approach
        ## refer to: Kenway et al. Effective adjoint approach for computational fluid dynamics,
        ## Progress in Aerospace Science, 2019.
        self.useAD = {"mode": "reverse", "dvName": "None", "seedIndex": -9999}
```
Can you please explain this bit?

**Comment by friedenhe** [comment]
Check here: https://dafoam.github.io/mydoc_verifications_overview.html

**Reply by uttamcadambi07** [reply]
Thank you. Very useful. So the seedindex gives the ith design variable right? In case of a design variable Field such as beta in field inversion, does the "seed" in the seedIndex used for forward mode AD that is printed at every iteration on the screen, correspond to the index at which the reverse mode AD stores the derivative in totalDerivative.txt file? What I mean to say is - if I want to compare the total derivative of a design variable at a particular index (in the state variable vector), I will set the seedIndex to be that value. Will I find the total derivative of the design variable at the same index in my totalDerivative.txt file when I run it in reverse mode?

**Reply by friedenhe** [reply]
> So the seedindex gives the ith design variable right?

Yes

> I will set the seedIndex to be that value. Will I find the total derivative of the design variable at the same index in my totalDerivative.txt file when I run it in reverse mode?

Yes

**Reply by uttamcadambi07** [reply]
Is the forwardAD implemented for FFD points? or is it only for field?

**Reply by friedenhe** [reply]
Yes, the forwardAD is implemented for FFD points. Check an example from here: https://github.com/DAFoam/verifications/blob/main/DARhoSimpleFoam/runScript.py
